Source/Diagnostics/ReducedDiags/DifferentialLuminosity.cpp

```diff
@@ -0,0 +1,272 @@
+
+            // Loop over cells
+            amrex::ParallelFor( n_cells,
+                [=] AMREX_GPU_DEVICE (int i_cell) noexcept
+            {
+                // The particles from species1 that are in the cell `i_cell` are
+                // given by the `indices_1[cell_start_1:cell_stop_1]`
+                index_type const cell_start_1 = cell_offsets_1[i_cell];
+                index_type const cell_stop_1  = cell_offsets_1[i_cell+1];
+                // Same for species 2
+                index_type const cell_start_2 = cell_offsets_2[i_cell];
+                index_type const cell_stop_2  = cell_offsets_2[i_cell+1];
+
+                for(index_type i_1=cell_start_1; i_1<cell_stop_1; ++i_1){
+                    for(index_type i_2=cell_start_2; i_2<cell_stop_2; ++i_2){
+
+                        index_type const j_1 = indices_1[i_1];
+                        index_type const j_2 = indices_2[i_2];
+
+                        Real const u1_square =  u1x[j_1]*u1x[j_1] + u1y[j_1]*u1y[j_1] + u1z[j_1]*u1z[j_1];
+                        Real const gamma1 = std::sqrt(1._rt + u1_square/c2);
+                        Real const u2_square = u2x[j_2]*u2x[j_2] + u2y[j_2]*u2y[j_2] + u2z[j_2]*u2z[j_2];
+                        Real const gamma2 = std::sqrt(1._rt + u2_square/c2);
+                        Real const u1_dot_u2 = u1x[j_1]*u2x[j_2] + u1y[j_1]*u2y[j_2] + u1z[j_1]*u2z[j_2];
+
+                        // center of mass energy
+                        Real const E_com = c2 * std::sqrt(m1*m1 + m2*m2 + 2*m1*m2* (gamma1*gamma2 - u1_dot_u2/c2));
+
+                        // determine particle bin
+                        int const bin = int(Math::floor((E_com-bin_min)/bin_size));
+
+                        if ( bin<0 || bin>=num_bins ) { continue; } // discard if out-of-range
+
+                        Real const v1_minus_v2_x = u1x[j_1]/gamma1 - u2x[j_2]/gamma2;
+                        Real const v1_minus_v2_y = u1y[j_1]/gamma1 - u2y[j_2]/gamma2;
+                        Real const v1_minus_v2_z = u1z[j_1]/gamma1 - u2z[j_2]/gamma2;
+                        Real const v1_minus_v2_square = v1_minus_v2_x*v1_minus_v2_x + v1_minus_v2_y*v1_minus_v2_y + v1_minus_v2_z*v1_minus_v2_z;
+
+                        Real const u1_cross_u2_x = u1y[j_1]*u2z[j_2] - u1z[j_1]*u2y[j_2];
+                        Real const u1_cross_u2_y = u1z[j_1]*u2x[j_2] - u1x[j_1]*u2z[j_2];
+                        Real const u1_cross_u2_z = u1x[j_1]*u2y[j_2] - u1y[j_1]*u2x[j_2];
+
+                        Real const v1_cross_v2_square = (u1_cross_u2_x*u1_cross_u2_x + u1_cross_u2_y*u1_cross_u2_y + u1_cross_u2_z*u1_cross_u2_z) / (gamma1*gamma1*gamma2*gamma2);
+
+                        Real const radicand = v1_minus_v2_square - v1_cross_v2_square / c2;
+
+                        Real const dL_dEcom = std::sqrt( radicand ) * w1[j_1] * w2[j_2] / dV / bin_size * dt; // m^-2 J^-1
+
+                        amrex::HostDevice::Atomic::Add(&dptr_data[bin], dL_dEcom);
```

Comment:
This works and is safe.

We can probably make it faster on CPUs for the OMP loop (where this now uses an `omp atomic update`) by using a copy of `d_data` per thread (block/tile) and reducing (sum) the copies after the OMP loop. @atmyers do you have a good construct for this? (I need something similar in ImpactX right now :D)

Reply:
The atomic-free version of this is implemented for amrex::DenseBins here: https://github.com/AMReX-Codes/amrex/blob/development/Src/Particle/AMReX_DenseBins.H#L342

Maybe with a little refactoring we could reuse this here?

Reply:
Yes, and we could maybe also use it here for the 1D charge binning:
https://github.com/ECP-WarpX/impactx/pull/627